### ISSUE TITLE
Wrong constant name with undefined constant error

### ISSUE THREAD

**Issue by jwoertink** [crystal-lang/crystal#7675]
When you're dealing with namespaces, things can be a little confusing. With [this example](https://play.crystal-lang.org/#/r/6qb9), you'll see that I have a `Tacos::Serializer` constant clearly defined. When you try to compile this program, you'll get a `undefined constant Tacos::Serializer` error, and the program doesn't compile. 
```crystal
class Tacos::Serializer
  def render
    {meat: "chicken"}
  end
end

class Api::Tacos::Index
  def call
    Tacos::Serializer.new.render
  end
end

route = Api::Tacos::Index.new
```

The reason it doesn't compile is because it's actually in the wrong namespace, so really, it's `Api::Tacos::Serializer` that's undefined, but the error message doesn't tell you this. That seems like a bug to me.

Also related to https://github.com/crystal-lang/crystal/issues/7554 which is requesting better error messages, I think it would be great if this error could say something like (but not necessarily) this

```
in src/actions/api/tacos/index.cr:4: undefined constant Tacos::Serializer in Api

    Tacos::Serializer.new
     ^~~~~~~~~~~~~~~
    Did you mean ::Tacos::Serializer?
```

Ref: 
https://github.com/luckyframework/lucky/issues/746

**Comment by paulcsmith** [comment]
I've been bitten by this bug a few times. The error displays the constant I typed but not the one Crystal tried to resolve. This leads to a wild goose chase trying to find typos or wondering why the class isn't required, when in reality it is because Crystal was looking in another namespace.

**Comment by jwoertink** [comment]
Just for my amusement, I tried that example in Ruby. As is, it throws a `uninitialized constant Tacos (NameError)`. Adding a `module Tacos; end` will throw `uninitialized constant Api (NameError)`. So to get it to run in Ruby, I needed this:

```ruby
module Tacos
end

module Api
  module Tacos
  end
end
class Tacos::Serializer
  def render
    {meat: "chicken"}
  end
end

class Api::Tacos::Index
  def call
    Tacos::Serializer.new.render
  end
end

route = Api::Tacos::Index.new
puts route.call
```

Which [still fails](https://play.crystal-lang.org/#/r/6qbg) in crystal.

**Comment by asterite** [comment]
The difference with Ruby is that I think Ruby considers `module Foo; class Bar` and `class Foo::Bar` as different things (in terms of namespace resolution inside that), which I find really confusing.

I'll see if I can improve the error message here (I wanted to improve the error messages in general but I don't think I'll have time for that).

**Comment by SlayerShadow** [comment]
Why don't you try to call an external namespace if it is needed?
```crystal
class Tacos::Serializer
  def render
    {meat: "chicken"}
  end
end

class Api::Tacos::Index
  def call
    ::Tacos::Serializer.new.render
  end
end
```
Works as expected. ( https://play.crystal-lang.org/#/r/6qib )
Constants resolution works different in Crystal and Ruby.

And sometimes Crystal resolves constants better:
https://play.crystal-lang.org/#/r/6qid - this does'n work in Ruby.

**Comment by paulcsmith** [comment]
@SlayerShadow Yes that works! :D The problem though is that the error message is wrong. Crystal says undefined constant `Tacos::Serializer` when it should say undefined constant `Api::Tacos::Serializer` because `Tacos::Serializer` *is* defined.

Also I love that we're all talking about a taco serializer :P

**Comment by asterite** [comment]
If somebody wants to tackle this, go ahead. I won't have time...

**Comment by HertzDevil** [comment]
The problem is `Tacos::Serializer` in that context could refer to any of the following:

* `::Api::Tacos::Index::Tacos::Serializer`
* `::Api::Tacos::Tacos::Serializer`
* `::Api::Tacos::Serializer`
* `::Tacos::Serializer`

whichever is defined first. So we cannot simply replace `Tacos::Serializer` in the error message with a single fully qualified name.
